AG-11370 Canvas focus refactor spike

olegat opened the pull request:
https://ag-grid.atlassian.net/browse/AG-11370

olegat committed:
AG-11370 Add listeners on <canvas> instead of on root <div>

This make sense because the InteractionManager was written when we only
had 1 <canvas> element. But now have multiple different types of elements
like <input type="range">, <button> and <div> elements.

olegat committed:
AG-11370 Move tab-guards around <canvas> instead of domManager root

olegat commented:
Note: This seems to fix the tabbing regressions that we've been seeing. However, the change to the `BASE_DOM` in `domManager.ts` is breaking the positioning of the tooltip. I'll have a look at that.

olegat committed:
AG-11370 Move tab-guard logic back into DOMManager

The Scene object should not case about the structure of the DOM tree.
Also simplify the Scene constructor and setContainer() methods, and
move the initialisation logic into chartContext.ts; This gets rid of
of a hidden cyclic dependency (note the removal of DOMManagerLike
interace in scene.ts).

olegat committed:
AG-11370 Add listeners to canvas only for keyboard events

olegat committed:
AG-11370 Remove unused code after rebase

olegat committed:
AG-11370 Fix eslint error

olegat committed:
AG-11370 Remove DOMManager.getDocumentCanvas()

olegat committed:
AG-11370 Fix DOMManager.setTabIndex

olegat committed:
AG-11370 Fix 'ag-charts-tab-guard' styling

The `display: none` property makes the element non-focusable, even if the
tabIndex is set.

alantreadway commented on packages/ag-charts-community/src/scene/scene.ts:
```diff
@@ -59,18 +47,9 @@ export class Scene {
-        }
+    setContainer(container: HTMLElement) {
+        this.canvas.element.remove();
+        container.appendChild(this.canvas.element);
```
I'd prefer if `DOMManager` was still responsible for creation + destruction of the canvas element.

olegat replied:
Do we still have a use-case for `Scene.setContainer` (e.g. integrated charts)?

The `Scene.domManager` is only used in the constructor, so the coordination between the constructor and `setContainer` doesn't make a lot of sense. It would be cleaner if the `Scene` constructor just took a non-nullable and `HTMLCanvasElement`, and leave it up to the caller to add that element in the correct place.

*Edit:* It seems like we use `setContainer` in sparklines. We could remove this usage to simply things (https://github.com/ag-grid/ag-grid/pull/8078). Or we could just not use the `setContainer` method in the ag-charts and add a `// comment` to clarify that this method is only intended to be used by sparklines.

alantreadway replied:
So yes the Integrated Charts case is real and exists - without this change, switching chart types yield a janky experience because we remove the old Chart instance and it's Canvas, then there maybe several frames before a new canvas is put back in place and the Chart rendered.

olegat replied:
Done. Change reverted.

alantreadway commented on packages/ag-charts-community/src/chart/interaction/interactionManager.ts:
```diff
@@ -159,7 +155,8 @@ export class InteractionManager extends BaseManager<InteractionTypes, Interactio
         private readonly keyboardOptions: { readonly enabled: boolean },
-        private readonly domManager: DOMManager
+        private readonly domManager: DOMManager,
+        private readonly canvasElement: HTMLCanvasElement
```
I'd prefer we didn't pass the `canvasElement` in here as it's implicitly hardcoding the expected DOM structure and relations within that here.

Can't we have the `DOMManager` understand where the appropriate places are to attach listeners to the DOM structure its responsible for?

olegat replied:
ACK. It should be feasible to update the logic of `DOMManager.addEventListener`

olegat replied:
✅ Done.
⚠️ Note: I had to tweak the logic in `DOMManager.addEventListener` a little bit to avoid firing the keyboard events twice.

alantreadway commented on packages/ag-charts-community/src/chart/chartContext.ts:
```diff
@@ -80,9 +80,13 @@ export class ChartContext implements ModuleContext {
         this.syncManager = syncManager;
         this.zoomManager = chart.zoomManager;
         this.domManager = new DOMManager(container);
-        scene?.setContainer(this.domManager);
```
The code here is slightly convoluted due to the 're-use resources from a destroyed chart instance' case, but I don't think we should be dealing with the raw elements here. The code here shouldn't care about specific elements are involved in the composition of the chart.

olegat replied:
> The code here shouldn't care about specific elements are involved in the composition of the chart.
Neither should the `Scene` object - it only needs an `HTMLCanvasElement`. I think that this is a good opportunity to this tidy up.

The initialisation of the chart-context shouldn't be so complex. The  `'re-use resources from a destroyed chart instance'` case isn't any different to a create-or-update logic.

olegat replied:
Done. Change reverted.

olegat committed:
AG-11370 CR: Use DOM.addEventListener in InteractionManager

olegat committed:
AG-11370 CR: Tidy up Scene initialisation

Avoid using Scene.setContainer. Instead, let the DOMManager handle the
DOM structuring.

alantreadway commented on packages/ag-charts-community/src/chart/dom/domManager.ts:
```diff
@@ -222,46 +218,53 @@ export class DOMManager extends BaseManager<Events['type'], Events> {
         listener: (this: HTMLElement, ev: HTMLElementEventMap[K]) => any,
         options?: boolean | AddEventListenerOptions
     ) {
-        this.element.addEventListener(type, listener, options);
```
Can we leave this as a general-purpose mechanism please, there is no need to hardcode the behaviour to `canvas` cases?

alantreadway commented on packages/ag-charts-community/src/chart/interaction/interactionManager.ts:
```diff
@@ -142,7 +142,6 @@ export class InteractionManager extends BaseManager<InteractionTypes, Interactio
     private readonly debug = Debug.create(true, 'interaction');
 
     private rootElement: HTMLElement;
-
```
Nit: Revert unecessary change please.

olegat committed:
AG-11370 CR: Revert most changes to DOMManager, Scene and ChartContext

olegat committed:
AG-11370 Fix build

alantreadway reviewed (approved):
Thanks for persevering on this one - LGTM!
